core/src/main/java/com/google/zxing/common/ECIStringBuilder.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2022 ZXing authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.zxing.common;
+
+import com.google.zxing.FormatException;
+
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Class that converts a sequence of ECIs and bytes into a string
+ *
+ * @author Alex Geller
+ */
+public final class ECIStringBuilder {
+  private StringBuilder currentBytes;
+  private StringBuilder result;
+  private Charset currentCharset = StandardCharsets.ISO_8859_1;
+
+  public ECIStringBuilder() {
+    currentBytes = new StringBuilder();
+  }
+  public ECIStringBuilder(int initialCapacity) {
+    currentBytes = new StringBuilder(initialCapacity);
+  }
+
+  /**
+   * appends {@code value} as a byte value
+   */
+  public void append(char value) {
+    currentBytes.append((char) (value & 0xff));
+  }
+
+  /**
+   * appends {@code value} as a byte value (not it's string representation)
+   */
+  public void append(byte value) {
+    currentBytes.append((char) (value & 0xff));
+  }
+
+  /**
+   * appends the characters in {@code value} as bytes values
+   */
+  public void append(String value) {
```

Comment:
Internally, this isn't tracking this as bytes, but as a String. That's OK if that's the correct semantics but it seems at odds with the docs. At the least isn't it true that it will ultimately use the ISO-8859-1 representation of the strings? you could build up a byte array directly instead (ByteArrayOutputStream)

Reply:
The outside semantics of the append methods is clearly bytes. But this class uses a StringBuilder as internal representation (and not, for example a ByteArray) because is used from the data matrix and the PDF417 decoders who both originally used StringBuilder (before the multi ECI support was added)  to decode the bytes. If we now use ByteArray here then we will degrade the performance because we will need a conversion from ByteArray to String at some point. The way that this class is written now, it has the same performance as the original StringBuilder in the case that there is no ECI (which is the most common case). 
The code that I have looked into in ZXing (code-128, qr-code, data-matrix and pdf417) uses Strings everywhere but in a large part of the code it is encoding and decoding bytes because these strings are assumed to not contain values greater 255. The differentiation may not have been important but this class now bridges between strings that only contain ISO-8859-1 characters (I call them bytes) and Strings that contain Unicode.  
By the way, I haven't looked at the details of issue #1163 but I suspect a missing check that would reject characters or trim Unicode characters to ISO-8859-1. Without this check or trim step, the PDF417 encoder assumes that the input contains only is ISO-8859-1 characters and it can for example produce an ArrayIndexOutOfBoundsException as described in the bug when a larger value is encountered.

Reply:
I think I'm generally fine staying in Strings in this process, but to your comment, this class already round-trips from String -> byte[] -> String, so seems like fewer hops, not more. I am worried about conflating strings and byte[]s back and forth - recipe for character-encoding-related problems. More direct seems better to reason about. If it's for convenience -- caller already makes the same encoding assumption -- I could see it. Just don't want to use string (assumed to be ISO-8859-1) a replacement for byte[] as it's so easy to work in byte[]s where they're bytes.

Reply:
There are fewer hops if we leave the code as is. The situation is as follows: The original code of both the data-matrix decoder and the PDF417 decoder declared "result" as StringBuilder() and appended ISO-8859-1 characters using the various StringBuilder append() methods. 
I then replaced "StringBuilder result=new StringBuilder()" with "ECIStringBuilder result=new ECIStringBuilder()" and the code remained largely the same since ECIStringBuilder provides the same append() methods. If the method ECIStringBuilder.appendECI() is not called during decoding then exactly on StringBuilder will have been constructed during the process and the final string will be obtained from it via result.toString() as in the original code before ECIs were introduced. There is never a conversion to bytes involved anywhere. Now, when result.appendECI() is called, then the ISO-8859-1 "characters" following the ECI are bytes of  the ECI that precedes them and will be converted to characters either on the next call to appendECI() or at latest when toString() is called. Here we have a character-byte-character conversion that could perhaps be changed but doing that without slowing the general case with no ECI will not be straightforward. The current code also avoids the byte-character conversion when appendECI() is called with a ISO-8859-1 ECI after if has previously been set to another encoding. This would not be possible if bytes were stored.